### DISCUSSION TITLE
Handle duplicate names in "Contributors" list

### DISCUSSION THREAD

**Discussion by lukemassa** [vuepress/core#1336] [Ideas]
#### What problem does this feature solve?

Right now, if there are two git commits for the same with the same `name` but different `email`, it will show up as two different "contributors". It is relatively common for the same person to commit with different emails, for example this instance from https://github.com/runatlantis/atlantis.
```
atlantis % git log runatlantis.io/docs/README.md | grep Author
Author: Luke Kysow <1034429+lkysow@users.noreply.github.com>
Author: Luke Kysow <1034429+lkysow@users.noreply.github.com>
Author: Luke Kysow <lkysow@gmail.com>
Author: Luke Kysow <lkysow@gmail.com>
Author: Luke Kysow <lkysow@gmail.com>
Author: Luke Kysow <lkysow@gmail.com>
Author: Luke Kysow <lkysow@gmail.com>
Author: Luke Kysow <lkysow@gmail.com>
```

This results in https://www.runatlantis.io/docs/ saying "Contributors: Luke Kysow, Luke Kysow" at the bottom.

<img width="326" alt="rendered" src="https://github.com/vuejs/vuepress/assets/2678195/a8570adf-59e4-4ba5-80de-589ee0795e76">

#### What does the proposed API look like?

Not sure, see below. 

#### How should this be implemented in your opinion?

I'm actually not sure. The issue is that these are rendered as `span`s, which are in fact different (see attached), but only in metadata not in any way visible on the page.

<img width="710" alt="html" src="https://github.com/vuejs/vuepress/assets/2678195/08f80cd8-41c8-449a-9996-21b04ae7ded1">

Here are a few ideas:
1. The simplest answer is "These are technically two different contributors, the code is working as expected". If this is the decision, that's fine, feel free to close this issue.
2. The deduping could happen at the git plugin level here: https://github.com/vuepress/vuepress-next/blob/9f21432e578c61fedfd4acc3466a08a58a0976a0/ecosystem/plugin-git/src/node/utils/getContributors.ts#L4, by adding a `filter()` and eliminating entries with the same `name` (picking I guess the first `email`?)
3. It could also be done at rendering time here: https://github.com/vuepress/vuepress-next/blob/9f21432e578c61fedfd4acc3466a08a58a0976a0/ecosystem/theme-default/src/client/components/PageMeta.vue#L115
4. Some sort of configuration option that allows users to either toggle this behavior, or to provide code that said "these two users are the same" or "run this filter on contributors". This seems kind of heavy-handed, and might not useful enough to belong in a public configuration, but 🤷 

#### Are you willing to work on this yourself?

Yup!

**Comment by Mister-Hope** [comment]
Nice catch, actually I am not sure what's the best too.

I think it's reasonable for the git plugin to output as is, and maybe the theme should filter duplicate names before displaying it?


Edited: for the git plugin, maybe it should try to drop any email containing no-reply if a contributor with same name already exisits.

**Reply by lukemassa** [reply]
Yeah I like your suggestion: basically omit anything that's obviously a duplicate, but nothing else. That way in the (rare) case that two people with the same name but different emails we include them both. Then later at the theme stage theoretically something could be done to distinguish them, but that could be a later improvement if it happens frequently.

It also opens the door to later refining what it means to be "obviously a duplicate"; this is the only specific case I know about now, so best to start conservatively, and users who see this problem will have a clear place to extend the code.

**Comment by Mister-Hope** [comment]
@meteorlxy Any suggestions?

**Comment by meteorlxy** [comment]
Maybe we can add an option for git plugin to manipulate the contributors list

**Comment by lukemassa** [comment]
@Mister-Hope this what you were thinking? https://github.com/lukemassa/vuepress-next/commit/c327018dc28c001f68385e695f8bb4ca63706b16 If so I'll turn this into an "issue" and write up a PR.

As for an option to manipulate the contributors list, I'm happy to try that out. I'm just not sure what else people might want to add other than this, and it might be kind of awkward to have people copy and paste this particular filter around. I recognize this represents a change in functionality over the existing situation, but it seems innocuous and unlikely to cause regressions.

**Reply by Mister-Hope** [reply]
I think we should widen the email to just contain `noreply` or `no-reply`

**Reply by lukemassa** [reply]
I incorporated this, and turned it into a PR https://github.com/vuepress/vuepress-next/pull/1364
